rfc/rfc-65/rfc-65.md

```diff
@@ -0,0 +1,163 @@
+
+- TTL policy definition & storage
+- Gathering proper partition statistics for TTL management
+- Executing TTL management
+
+### TTL Policy Definition
+
+We have four main considerations when designing TTL policy:
+
+1. Usually user hopes to manage partition TTL by expired time, so we support the policy `KEEP_BY_TIME` that partitions
+   will expire N days after their last modified time. Note that the last modified time here means any inserts/updates to
+   the partition. Also, we will make it easy to extend new policies so that users can implement their own polices to
+   decide which partitions are expired.
+2. User need to set different policies for different partitions. For example, the hudi table is partitioned
+   by `user_id`. For partition(user_id='1'), we set the policy that this partition will expire 7 days after its last
+   modified time, while for
+   partition(user_id='2') we can set the policy that it will expire 30 days after its last modified time.
+3. It's possible that users have multi-fields partitioning, and they don't want set policy for all partition. So we need
+   to support partition regex when defining
+   partition TTL policies. For example, for a hudi table partitioned by (user_id, ts), we can first set a default
+   policy(which policy value contains `*`)
+   that for all partitions matched (user_id=*/) will expire in 7 days, and explicitly set a policy that for partition
+   matched user_id=3 will expire in 30 days. Explicit policy will override default polices, which means default policies
+   only takes effects on partitions that do not match any explicit ttl policies.
+4. We may need to support record level ttl policy in the future. So partition TTL policy should be an implementation of
+   HoodieTTLPolicy.
+
+So we have the TTL policy definition like this (may change when implementing):
+
+```java
+public class HoodiePartitionTTLPolicy implements HoodieTTLPolicy {
+  public enum TTLPolicy {
+    // We supports keep by last modified time at the first version
+    KEEP_BY_TIME
+  }
+
+  // Partition spec for which the policy takes effect, could be a regex or a static partition value
+  private String partitionSpec;
+
+  private TTLPolicy policy;
+
+  private long policyValue;
+}
+```
+
+### User Interface for TTL policy Definition
+
+Users can config partition TTL management policies through SparkSQL Call Command and through table config directly.
+Assume that the user has a hudi table partitioned by two fields(user_id, ts), he can config partition TTL policies as
+follows.
+
+```sparksql
+-- Set policy for all user_id using partition regex, which keeps the data for 30 days.
+call add_ttl_policy(table => 'test', partitionSpec => 'user_id=*/', policy => 'KEEP_BY_TIME', policyValue => '30');
+
+--For partition user_id=3/, keep the data for 7 day.
+call add_ttl_policy(table => 'test', partitionSpec => 'user_id=3/', policy => 'KEEP_BY_TIME', policyValue => '7');
```

Comment:
If adding different ttl policy on the same partition, which ttl policy would valid?